http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/AutoDeleteAppMonitor.scala

```diff
@@ -0,0 +1,134 @@
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage.DeleteAppMessage
+import org.broadinstitute.dsde.workbench.leonardo.{
+  AllowedChartName,
+  AppId,
+  AppName,
+  AppStatus,
+  CloudContext,
+  SamResourceId
+}
+import org.broadinstitute.dsde.workbench.model.{TraceId, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.broadinstitute.dsp.ChartName
+import org.typelevel.log4cats.StructuredLogger
+
+import java.time.Instant
+import scala.concurrent.ExecutionContext
+
+/**
+ * This monitor periodically sweeps the Leo database and autodelete App that have been running for too long.
+ */
+class AutoDeleteAppMonitor[F[_]](
+  config: AutoDeleteConfig,
+  publisherQueue: Queue[F, LeoPubsubMessage],
+  authProvider: LeoAuthProvider[F]
+)(implicit
+  dbRef: DbReference[F],
+  logger: StructuredLogger[F],
+  ec: ExecutionContext,
+  openTelemetry: OpenTelemetryMetrics[F]
+) extends BackgroundProcess[F, AppToAutoDelete] {
+  override def name: String =
+    "autodeleteApp" //
+  override def interval: scala.concurrent.duration.FiniteDuration = config.autodeleteCheckInterval
+
+  override def getCandidates(now: Instant)(implicit
+    F: Async[F],
+    metrics: OpenTelemetryMetrics[F],
+    logger: StructuredLogger[F]
+  ): F[Seq[AppToAutoDelete]] = for {
+    candidates <- appQuery.getAppsReadyToAutoDelete.transaction
+  } yield candidates
+
+  override def action(a: AppToAutoDelete, traceId: TraceId, now: Instant)(implicit F: Async[F]): F[Unit] =
+    for {
+      now <- F.realTimeInstant
+      loggingCtx = Map("traceId" -> traceId.asString)
+      _ <- a.cloudContext match {
+        case CloudContext.Gcp(googleProject) =>
+          if (a.appStatus == AppStatus.Error) {
+            implicit val implicitTraceId: Ask[F, TraceId] = Ask.const[F, TraceId](traceId)
+            for {
+              // we only need to delete Sam record for clusters in Google. Sam record for Azure is managed by WSM
+              _ <- authProvider.notifyResourceDeleted(
+                a.samResourceId,
+                a.creator,
+                googleProject
+              )
+              _ <- appQuery.markAsDeleted(a.id, now).transaction
+            } yield ()
+          } else {
+            for {
+              _ <- KubernetesServiceDbQueries.markPreDeleting(a.id).transaction
+              deleteMessage = DeleteAppMessage(
+                a.id,
+                a.appName,
+                googleProject,
+                None,
+                Some(traceId)
+              )
+              trackUsage = AllowedChartName.fromChartName(a.chartName).exists(_.trackUsage)
+              _ <- appUsageQuery.recordStop(a.id, now).whenA(trackUsage).recoverWith { case e: FailToRecordStoptime =>
+                logger.error(loggingCtx)(e.getMessage)
+              }
+              _ <- publisherQueue.offer(deleteMessage)
```

Comment:
It looks like in the `else` clause we are taking into consideration the `app` is in an error state.

I am curious why the two approaches are here for GCP based on `a.appStatus == AppStatus.Error` versus (potentially) being in how we respond to the `DeleteAppMessage`.

Some other questions:
- is this meant to replicate some of the logic we are doing if a user explicitly requests to delete an app, and 
- if so, is it possible to refactor code so that we can be sure to be using the same logic via both code paths?

Reply:
I do not have deep Leo knowledge, and I am not the original author of deleteApp call. My understanding is there is a separate cron job that translate ERROR app into DELETED? Better to check with IA folk thought.. 
I agree we can refactor code into a common place, but I would need more help & guidance on this if you want that. I have very little scala knowledge and Leo codebase, and may have different best practice principles in my mind. If you want me to do the refactor, can you provide me: method name, file name to put it? Thanks

Reply:
I agree with the general sentiment, but a refactoring would be outside of the scope of this PR IMO (also refactoring this would create a ton of merge conflict with the deleteApp endpoints I am adding in these files :) )